src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AQNBuilder.Mono.cs

```diff
@@ -0,0 +1,483 @@
+
+				Append("[]");
+			}
+			public void AddArray(int rank)
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR))
+					Fail();
+
+				m_parseState = ParseState.PTRARR;
+
+				if (rank <= 0)
+					throw new ArgumentOutOfRangeException ();
+
+				if (rank == 1)
+					Append("[*]");
+				else if (rank > 64)
+				{
+					// Only taken in an error path, runtime will not load arrays of more than 32 dimentions
+					Append ($"[{rank}]");
+				}
+				else
+				{
+					var strDim = new StringBuilder (rank + 3);
+
+					strDim.Append ('[');
+					for(int i = 1; i < rank; i++) strDim.Append (',');
+					strDim.Append (']');
+					Append(strDim.ToString ());
+				}
+			}
+			public void AddAssemblySpec(string assemblySpec)
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR | ParseState.BYREF))
+					Fail();
+
+				m_parseState = ParseState.ASSEMSPEC;
+
+				if (assemblySpec != null && !assemblySpec.Equals(String.Empty))
+				{
+					Append(", ");
+
+					if (m_instNesting > 0)
+					{
+						EscapeEmbeddedAssemblyName(assemblySpec);
+					}
+					else
+					{
+						EscapeAssemblyName(assemblySpec);
+					}
+
+					m_bHasAssemblySpec = true;
+				}
+
+			}
+			public void ToString(out string retString)
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR | ParseState.BYREF | ParseState.ASSEMSPEC))
+					Fail();
+
+				if (m_instNesting == 0)
+					Fail();
+				retString = m_pStr.ToString ();
+			}
+
+			public void Clear()
+			{
+				if (m_pStr != null)
+				{
+					m_pStr.Clear ();
+				}
+				m_bNestedName = false;
+				m_instNesting = 0;
+				m_bFirstInstArg = false;
+				m_parseState = ParseState.START;
+				m_bHasAssemblySpec = false;
+				m_bUseAngleBracketsForGenerics = false;
+				// m_stack.Clear (); looks like Mono's Stack type is missing Clear()
+				m_stack = new Stack ();
+			}
+
+			private void EscapeName (string name)
+			{
+				if (TypeString.ContainsReservedChar (name))
+				{
+					foreach (char c in name.AsSpan ())
+					{
+						if (c == '\0')
+							break;
+						if (TypeString.IsTypeNameReservedChar (c))
+							m_pStr.Append ('\\');
+						m_pStr.Append (c);
+					}
+				}
+				else
+					Append (name);
+			}
+
+			private void EscapeAssemblyName (string name)
+			{
+				Append (name);
+			}
+
+			private void EscapeEmbeddedAssemblyName (string name)
+			{
+				bool bContainsReservedChar = false;
+
+				foreach (char c in name.AsSpan ())
+				{
+					if (c == ']')
+					{
+						bContainsReservedChar = true;
+						break;
+					}
+				}
+
+				if (bContainsReservedChar)
+				{
+					foreach (char c in name.AsSpan ())
+					{
+						if (c == ']')
+							Append ('\\');
+
+						Append (c);
+					}
+				}
+				else
+				{
+					Append (name);
+				}
+
+			}
+
+			private bool CheckParseState (ParseState validState) { return (m_parseState & validState) != 0;}
+			private void Fail () { m_parseState = ParseState.ERROR; }
```

Comment:
The curly brackets formatting is mix of everything, it would look nicer to format them consistently